### ISSUE TITLE
sbcl 2.4.9 fails to compile random-state, SHOP3 can't be built

### ISSUE THREAD

**Issue by ko56** [shop-planner/shop3#169]
So SHOP3 fails to build.

```
cl-user> (require "mhp")
; compiling file "/home/ko/build/shop3/jenkins/ext/random-state/quasi.lisp" 
(written 08 AUG 2024 09:09:38 AM):

; file: /home/ko/build/shop3/jenkins/ext/random-state/quasi.lisp
; in: define-generator quasi
;     (SETF LAST 0.0d0)
; --> SB-KERNEL:THE* 
; ==>
;   0.0d0
; 
; caught warning:
;   Constant 0.0 conflicts with its asserted type single-float.
;   See also:
;     The SBCL Manual, Node "Handling of Types"

;     (RANDOM-STATE:RANDOM-UNIT RANDOM-STATE::SOURCE)
; 
; note: forced to do full call
;       unable to do inline float coercion (cost 5) because:
;       The first argument is a (integer 4611686018427387904), not a (signed-byte
;                                                                     64).

;     (MOD (+ LAST 0.5d0 (RANDOM-STATE:RANDOM-UNIT RANDOM-STATE::SOURCE)) 1.0d0)
; 
; caught warning:
;   Derived type of (- number (* (+ sb-c::quot 0.0))) is
;     (values double-float &optional),
;   conflicting with its asserted type
;     single-float.
;   See also:
;     The SBCL Manual, Node "Handling of Types"
; 
; note: doing float to pointer coercion (cost 13), for:
;       the first argument of cons


; wrote /home/ko/.cache/common-lisp/sbcl-2.4.9-bsd-x64/home/ko/build/shop3/jenkins/ext/random-state/quasi-tmp5GEXGEG5.fasl
; compilation finished in 0:00:00.017
ASDF could not load mhp because
compile-file-error while compiling #<cl-source-file "random-state" "quasi">.

```

**Comment by rpgoldman** [comment]
Thanks for letting me know!  I will push a fix as soon as I can. Likely this needs a refresh of the bundled version of `random-state`.

It would be ideal if I could figure out how to use `qlot` or something similar to make it possible to use quicklisp for the dependencies, instead of git submodules...

**Comment by rpgoldman** [comment]
I just built SBCL 2.4.9 on my Mac and was able to build SHOP3 successfully.  This is on ARM64, so I suppose that could account for a difference in float representation?

What is `*read-default-float-format*` in your SBCL?  Is there any chance that it's `'double-float`? If so, that might account for the error you are seeing.  Looks like the error is here:

```
(define-generator quasi 'single-float (stateful-generator)
    ((source *generator* :type T)
     (last 0.0 :type single-float))
  (:reseed
   (reseed source seed)
   (setf last 0.0))
  (:next
   (declare (optimize speed (safety 1)))
   (setf last (mod (+ last 0.5 (random-unit source)) 1.0))))
   ```
   
You should probably also report this [upstream](https://github.com/Shinmera/random-state).

If you want a quick fix, and for some reason `*read-default-float-format*` in your SBCL defaults to double float (it shouldn't -- the ANSI spec defines it as defaulting to single float), then you could replace 0.0 with 0.0f0 and the same for 0.5 and 1.0 -- the `f` specifies single float.

You should also consider looking for code in your lisp image that *sets* `*read-default-float-format*` instead of *binding* it, and change it so that code doesn't leak float read configuration out into its environment.

If this takes care of things for you, please close this issue.

**Comment by ko56** [comment]
Thanks for the (quick!) diagnosis. Yes, I have `*read-default-float-format*` set to double float in my `~/.sbclrc.`  I made this change a little while ago, because it seemed to be changing `double-float-epsilon`. But I just re-tested, and it's not affecting it now.
My quick fix was to revert to sbcl 2.4.5, which works fine with this setting of `*read-default-float-format*`.
I now removed the setting from `~/.sbclrc`, changed back to sbcl 2.4.9, and everything works.
 Thanks very much.
